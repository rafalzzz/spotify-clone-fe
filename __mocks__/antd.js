const antd = jest.requireActual('antd');

module.exports = {
  ...antd,
};
