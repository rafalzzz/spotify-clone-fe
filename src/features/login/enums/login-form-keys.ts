export enum LoginFormKeys {
  LOGIN = 'login',
  PASSWORD = 'password',
  REMEMBER_ME = 'rememberMe',
}
