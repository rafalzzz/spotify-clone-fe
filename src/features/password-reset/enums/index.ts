export { PasswordResetFormKeys } from './password-reset-form-keys';
export { PasswordResetCompleteFormKeys } from './password-reset-complete-form-keys';
