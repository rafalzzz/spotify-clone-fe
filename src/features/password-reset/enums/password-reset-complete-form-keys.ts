export enum PasswordResetCompleteFormKeys {
  NEW_PASSWORD = 'newPassword',
  REPEAT_NEW_PASSWORD = 'repeatNewPassword',
}
