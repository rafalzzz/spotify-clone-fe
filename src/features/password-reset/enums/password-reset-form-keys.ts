export enum PasswordResetFormKeys {
  LOGIN = 'login',
}
