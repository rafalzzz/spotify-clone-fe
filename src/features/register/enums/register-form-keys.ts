export enum RegisterFormKeys {
  EMAIL = 'email',
  PASSWORD = 'password',
  NICKNAME = 'nickname',
  DAY = 'day',
  MONTH = 'month',
  YEAR = 'year',
  GENDER = 'gender',
  OFFERS = 'offers',
  SHARE_INFORMATION = 'shareInformation',
  TERMS = 'terms',
}
