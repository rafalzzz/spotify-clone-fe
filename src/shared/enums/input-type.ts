export enum InputType {
  TEXT = 'text',
  PASSWORD = 'password',
  SELECT = 'select',
  RADIO = 'radio',
  CHECKBOX = 'checkbox',
  SWITCH = 'switch',
}

export enum NonStandardInputType {
  DATE_OF_BIRTH = 'dateOfBirth',
}
