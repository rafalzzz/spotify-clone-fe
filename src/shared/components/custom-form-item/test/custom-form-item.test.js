import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as antdOriginal from 'antd';
import { Form } from 'antd';
import React from 'react';

import { InputType } from '@/enums/input-type';

import { CustomFormItem } from '..';

import '@testing-library/jest-dom';

const textInputPlaceholder = 'Enter text';
const inputValidator = jest.fn();
const mockSetFieldValue = jest.fn();

const sharedFormItemProps = {
  key: 'text',
  name: 'text',
  label: 'Enter text',
  rules: [{ validator: inputValidator }],
};

const textInputProps = {
  ...sharedFormItemProps,
  inputProps: {
    placeholder: textInputPlaceholder,
  },
};

const MOCKED_OPTIONS = [
  { label: 'label1', value: 'value1' },
  { label: 'label2', value: 'value2' },
];

const MOCKED_FORM_ITEMS = [
  {
    type: InputType.TEXT,
    props: textInputProps,
  },
  {
    type: InputType.PASSWORD,
    props: textInputProps,
  },
  {
    type: InputType.SELECT,
    props: {
      ...sharedFormItemProps,
      selectProps: {
        options: MOCKED_OPTIONS,
      },
    },
  },
  {
    type: InputType.RADIO,
    props: {
      ...sharedFormItemProps,
      radioProps: {
        options: [
          { label: 'Test1', value: 'test1' },
          { label: 'Test2', value: 'test2' },
        ],
      },
    },
  },
  {
    type: InputType.CHECKBOX,
    props: {
      key: 'text',
      name: 'text',
      checkboxProps: {
        text: 'Text example',
      },
    },
  },
  {
    type: InputType.SWITCH,
    props: {
      key: 'text',
      name: 'text',
      switchProps: {
        label: 'Text example',
      },
    },
  },
];

const renderCustomFormItem = (type, formItemProps) => {
  const props = {
    type,
    ...formItemProps,
    setFieldValue: mockSetFieldValue,
  };

  return render(
    // It's necessary to warp CustomFormItem into Form for test purposes
    <Form>
      <CustomFormItem {...props} />
    </Form>,
  );
};

describe('CustomFormItem', () => {
  it('render component without error', () => {
    const { type, props } = MOCKED_FORM_ITEMS[0];
    const screen = renderCustomFormItem(type, props);
    expect(screen).toMatchSnapshot();
  });

  describe('render correct input type', () => {
    MOCKED_FORM_ITEMS.forEach(({ type, props }) => {
      it(type, () => {
        const { queryByTestId } = renderCustomFormItem(type, props);
        expect(queryByTestId(`input-type-${type}`)).toBeInTheDocument();
      });
    });
  });

  describe('error handling', () => {
    it('validates field when input value change', async () => {
      const mockedInputValue = 'test';
      inputValidator.mockImplementation((getFieldValue) => () => Promise.resolve());

      const { type, props } = MOCKED_FORM_ITEMS[0];
      const { getByPlaceholderText, getByText } = renderCustomFormItem(type, props);

      const textInput = getByPlaceholderText(textInputPlaceholder);
      expect(textInput).toBeInTheDocument();

      await userEvent.type(textInput, mockedInputValue);

      expect(textInput).toHaveValue(mockedInputValue);
      expect(inputValidator).toHaveBeenCalled();
    });

    it('throws error for invalid input type', () => {
      const invalidInputType = 'invalid_input_type';
      const { props } = MOCKED_FORM_ITEMS[0];

      jest.spyOn(console, 'error');
      console.error.mockImplementation(() => {});

      expect(() => {
        render(<CustomFormItem type={invalidInputType} {...props} />);
      }).toThrowError(`${invalidInputType} input type does not exist`);

      console.error.mockRestore();
    });
  });
});
