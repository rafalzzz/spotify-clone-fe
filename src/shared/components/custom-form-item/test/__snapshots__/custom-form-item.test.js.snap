// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomFormItem render component without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-horizontal css-dev-only-do-not-override-12upa3x"
      >
        <div
          class="ant-form-item form-item css-dev-only-do-not-override-12upa3x"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-12upa3x"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-12upa3x"
            >
              <label
                class=""
                for="text"
                title=""
              >
                <span
                  class="form-item__label"
                >
                  Enter text
                </span>
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-12upa3x"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input css-dev-only-do-not-override-12upa3x"
                    data-testid="input-type-text"
                    id="text"
                    placeholder="Enter text"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-12upa3x"
    >
      <div
        class="ant-form-item form-item css-dev-only-do-not-override-12upa3x"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-12upa3x"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-12upa3x"
          >
            <label
              class=""
              for="text"
              title=""
            >
              <span
                class="form-item__label"
              >
                Enter text
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-12upa3x"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input css-dev-only-do-not-override-12upa3x"
                  data-testid="input-type-text"
                  id="text"
                  placeholder="Enter text"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
